tool/tsh/common/tsh.go

```diff
@@ -3263,13 +3327,17 @@ func retryWithAccessRequest(
 	resource string,
 ) error {
 	origErr := fn()
-	if cf.disableAccessRequest || !trace.IsAccessDenied(origErr) {
-		// Return if --disable-access-request was specified.
+	if cf.RequestMode == accessRequestModeOff || !trace.IsAccessDenied(origErr) {
+		// Return if --request-mode=off was specified.
 		// Return the original error if it's not AccessDenied.
 		// Quit now if we don't have a hostname.
 		return trace.Wrap(origErr)
 	}
 
+	// Print and log the original AccessDenied error.
+	fmt.Fprintln(os.Stderr, utils.UserMessageFromError(origErr))
+	fmt.Fprintf(os.Stdout, "You do not currently have access to %q, attempting to request access.\n\n", resource)
+
```

Comment:
I disagree with moving these prints before the call to `onAccessRequestCreator` when the RequestMode is resource. If it returns an error we won't actually attempt to request access and we don't want users to see this for like, every failed `tsh ssh`.

But I do understand that you want to show this before prompting the user to select a role.

I might refactor this function to something like
```go
req, err := onAccessRequestCreator(cf, tc)
if err != nil {
	// We can't or shouldn't request access to the resource. Log
	// a short debug message in case this is unexpected, but return the
	// original AccessDenied error from the ssh attempt which is likely to
	// be far more relevant to the user.
	log.Debugf("Not attempting to automatically request access, reason: %v", err)
	return trace.Wrap(origErr)
}
req, err := promptUserForAccessRequestDetails(cf, req)
if err != nil {
	return trace.Wrap(err)
}
```

`onAccessRequestCreator` (which is often `accessRequestForSSH`) would return a request will all of the requestable roles populated. `promptUserForAccessRequestDetails` would handle printing the "You do not currently have access to %q, attempting to request access." string, prompting for a single role if doing a role request, and prompting for a request reason